runtime/src/accounts_index.rs

```diff
@@ -581,19 +883,79 @@ impl<T: 'static + Clone> AccountsIndex<T> {
+        // provides the ancestors to check. So even if a zero-lamport account is not yet
+        // removed from the secondary index, the scan function will:
+        // 1) consult the primary index via `get(&pubkey, Some(ancestors), max_root)`
+        // and find the zero-lamport version
+        // 2) When the fetch from storage occurs, it will return Account::Default
+        // (as persisted tombstone for snapshots). This will then ultimately be
+        // filtered out by post-scan filters, like in `get_filtered_spl_token_accounts_by_owner()`.
+        if *account_owner == inline_spl_token_v2_0::id()
+            && account_data.len() == inline_spl_token_v2_0::state::Account::get_packed_len()
+        {
+            if account_indexes.contains(&AccountIndex::SplTokenOwner) {
+                let owner_key = Pubkey::new(
+                    &account_data[SPL_TOKEN_ACCOUNT_OWNER_OFFSET
+                        ..SPL_TOKEN_ACCOUNT_OWNER_OFFSET + PUBKEY_BYTES],
+                );
+                self.spl_token_owner_index.insert(&owner_key, pubkey, slot);
+            }
+
+            if account_indexes.contains(&AccountIndex::SplTokenMint) {
+                let mint_key = Pubkey::new(
+                    &account_data[SPL_TOKEN_ACCOUNT_MINT_OFFSET
+                        ..SPL_TOKEN_ACCOUNT_MINT_OFFSET + PUBKEY_BYTES],
+                );
+                self.spl_token_mint_index.insert(&mint_key, pubkey, slot);
+            }
+        }
+    }
+
     // Updates the given pubkey at the given slot with the new account information.
     // Returns true if the pubkey was newly inserted into the index, otherwise, if the
     // pubkey updates an existing entry in the index, returns false.
     pub fn upsert(
         &self,
         slot: Slot,
         pubkey: &Pubkey,
+        account_owner: &Pubkey,
+        account_data: &[u8],
+        account_indexes: &HashSet<AccountIndex>,
         account_info: T,
         reclaims: &mut SlotList<T>,
     ) -> bool {
         let (mut w_account_entry, is_newly_inserted) =
             self.get_account_write_entry_else_create(pubkey);
+        // We don't atomically update both primary index and secondary index together.
+        // This certainly creates small time window with inconsistent state across the two indexes.
+        // However, this is acceptable because:
+        //
+        //  - A strict consistent view at any given moment of time is not necessary, because the only
+        //  use case for the secondary index is `scan`, and `scans` are only supported/require consistency
+        //  on frozen banks, and this inconsistency is only possible on working banks.
+        //
+        //  - The secondary index is never consulted as primary source of truth for gets/stores.
+        //  So, what the accounts_index sees alone is sufficient as a source of truth for other non-scan
+        //  account operations.
         w_account_entry.update(slot, account_info, reclaims);
```

Comment:
nits: maybe we can `drop(w_account_entry)`? I'd prefer `{...}` over `drop()` for readability and maintainability, though.

Reply:
hmm that's a good point, should be no need to hold the lock.